runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java

```diff
@@ -59,14 +59,25 @@ OutputT apply(PTransform<InputT,OutputT> transform, InputT input) {
   @Override
   public FlinkRunnerResult run(Pipeline pipeline) {
     try {
-      return delegate.run(pipeline);
-    } catch (RuntimeException e) {
+      FlinkRunnerResult result = delegate.run(pipeline);
+
+      return result;
+    } catch (Throwable e) {
       // Special case hack to pull out assertion errors from PAssert; instead there should
       // probably be a better story along the lines of UserCodeException.
-      if (e.getCause() != null
-          && e.getCause() instanceof JobExecutionException
-          && e.getCause().getCause() instanceof AssertionError) {
-          throw (AssertionError) e.getCause().getCause();
+      Throwable cause = e;
+      Throwable oldCause = e;
+      do {
+        if (cause.getCause() == null) {
+          break;
+        }
+
+        oldCause = cause;
+        cause = cause.getCause();
+
+      } while (!oldCause.equals(cause));
+      if (cause instanceof AssertionError) {
+        throw (AssertionError) cause;
       } else {
         throw e;
```

Comment:
What is the purpose of the AssertionError check? You throw the exception either way.